packages/network-controller/src/create-network-client.ts

```diff
@@ -194,6 +165,172 @@ export function createNetworkClient({
   return { configuration, provider, blockTracker, destroy };
 }
 
+/**
+ * Creates an RPC service chain, which represents the primary endpoint URL for
+ * the network as well as its failover URLs.
+ *
+ * @param args - The arguments.
+ * @param args.id - The ID that will be assigned to the new network client in
+ * the registry.
+ * @param args.primaryEndpointUrl - The primary endpoint URL.
+ * @param args.configuration - The network configuration.
+ * @param args.getRpcServiceOptions - Factory for constructing RPC service
+ * options. See {@link NetworkControllerOptions.getRpcServiceOptions}.
+ * @param args.messenger - The network controller messenger.
+ * @param args.isRpcFailoverEnabled - Whether or not requests sent to the
+ * primary RPC endpoint for this network should be automatically diverted to
+ * provided failover endpoints if the primary is unavailable. This effectively
+ * causes the `failoverRpcUrls` property of the network client configuration
+ * to be honored or ignored.
+ * @param args.logger - A `loglevel` logger.
+ * @returns The RPC service chain.
+ */
+function createRpcServiceChain({
+  id,
+  primaryEndpointUrl,
+  configuration,
+  getRpcServiceOptions,
+  messenger,
+  isRpcFailoverEnabled,
+  logger,
+}: {
+  id: NetworkClientId;
+  primaryEndpointUrl: string;
+  configuration: NetworkClientConfiguration;
+  getRpcServiceOptions: (
+    rpcEndpointUrl: string,
+  ) => Omit<RpcServiceOptions, 'failoverService' | 'endpointUrl'>;
+  messenger: NetworkControllerMessenger;
+  isRpcFailoverEnabled: boolean;
+  logger?: Logger;
+}) {
+  const availableEndpointUrls: [string, ...string[]] = isRpcFailoverEnabled
+    ? [primaryEndpointUrl, ...(configuration.failoverRpcUrls ?? [])]
+    : [primaryEndpointUrl];
+  const buildRpcServiceConfiguration = (endpointUrl: string) => ({
+    ...getRpcServiceOptions(endpointUrl),
+    endpointUrl,
+    logger,
+  });
```

Comment:
Nit: I don't think we need a builder function here, it's only called in one place. We can do something like

```suggestion
  const rpcServiceConfigurations = availableEndpointUrls.map((endpointUrl) => ({
    ...getRpcServiceOptions(endpointUrl),
    endpointUrl,
    logger,
  })),
```

Then reference `rpcServiceConfigurations[0]` and `...rpcServiceConfigurations.slice(1)` later in the constructor.

At least this seems simpler to me? At least slightly.